content/blog/2023/2023-06-25-archunit.md

```diff
@@ -0,0 +1,598 @@
+Method <io.reflectoring.archunit.api.EmployeeController.getEmployee()> calls constructor <io.reflectoring.archunit.persistence.EmployeeDao.<init>()> in (EmployeeController.java:15)
+Method <io.reflectoring.archunit.api.EmployeeController.getEmployee()> calls method <io.reflectoring.archunit.persistence.EmployeeDao.findEmployee()> in (EmployeeController.java:16)
+```
+
+The example shows how easy it is to use ArchUnit in a Java project. Before we look at more examples, let's discuss why
+architecture violations are typically introduced in projects over time. 
+
+## Reasons for Architecture Erosion Over Time
+There are many reasons why developers start to deviate from the initial design choices, coding best practices, or testing practices.
+One of the most common reasons is probably time pressure. As this is rather straightforward, we we'll look at some other reasons
+in more detail.
+
+### Architecture Awareness
+At the start of a software project, we usually take certain design choices and organize the code in methods, classes, 
+packages, modules, and layers. Each of these has its specific purpose and a clear boundary. 
+The data access layer for example should have the sole responsibility to retrieve persisted data.
+It should not provide an API endpoint or map data to an external format like JSON or XML. 
+
+We also make choices on certain implementation details like the inheritance (for example, every DAO class should implement
+an interface), or how to handle date and time in the code.
+
+Let's look at a simple example. Instead of:
+
+```java
+LocalDateTime localDate = LocalDateTime.now();
+```
+
+we might want to use:
+
+```java
+LocalDateTime localDate = LocalDateTime.now(clock);
+```
+
+When we decide to use the latter way of instantiating our object, we'll probably remember the reason for a while. However, after some
+time we might forget. Also, other developers who join the project might unintentionally deviate from the original choice. 
+
+With ArchUnit, we can add a test that will fail if the static factory method ```now``` is used without the parameter: 
+
+```java
+@Test
+public void alwayInjectClockToFaciliateTesting() { 
+  // ...
+}
+```
+
+Such a test reminds the developer to use the parameter and remain consistent within the codebase. It also explains the reason
+why to use (or not to use) a specific method.
+
+This is a good example of how we can use ArchUnit to document the intended architecture in the form of unit tests. Of course,
+we can change things when we see the need for it. There might be good reasons to deviate from a certain pattern. However,
+using tests as a documentation will remind us to think about why we want to deviate.
+
+## ArchUnit Examples 
+<a name="examples"></a>
+Most examples of how to use ArchUnit describe checks on dependencies between classes and packages.
+That's however not the only use case. Let's have a look at three other types of checks that we can create.
+
+### @deprecated
```

Comment:
Please modify the code to eliminate the need for horizontal scrolling by the reader in all code examples.